### PR TITLE
Work on FATE#332100: manual tickets (Booth, Geo)

### PR DESCRIPTION
**Pull request by tomschr**
This pull request contains the following changes:

* Add new section about normal and manual tickets (explaining some properties)
* Resort `calloutlist` as the manual mode lead to a incorrect numbering
* Describe manual mode in config file (id=co.ha.geo.booth.mode)
* Try to use "booth" instead of "booth client" as there is no "client" subcommand
* Document `-F` option when granting a ticket

### CODE REVIEW

File: xml/geo_booth_i.xml [hunk 1]
```diff
@@ -64,6 +64,73 @@
...
+     <para>The normal ticket management is not applied to manually controlled
+      tickets. Manual tickets do not have elections, automatic failover
+      procedures, and expirations.</para>
+    </formalpara>
+   </listitem>
+   <listitem>
+    <formalpara>
+     <title>Manual tickets are not moved</title>
+     <para>Tickets which were manually granted to a site, will remain there
+      until they are manually revoked. Even if a site becomes offline, the
```
Comment (taroth21) [inline xml/geo_booth_i.xml]: becomes offline -> goes offline? (or: Even if a site is no longer online, ...)

typo: tiket -> ticket

some services -> the services that depend on a ticket

File: xml/geo_booth_i.xml [hunk 2]
```diff
@@ -64,6 +64,73 @@
...
+      tiket will not be moved to another site. This behavior ensures that
+      some services remain on a particular site and are not moved to another
+      site.
+     </para>
+    </formalpara>
+   </listitem>
+   <listitem>
+    <formalpara>
+     <title>Same commands for manual tickets</title>
```
Comment (taroth21) [inline xml/geo_booth_i.xml]: Same commands for manual tickets -> Same commands for managing both types of tickets?

File: xml/geo_booth_i.xml [hunk 3]
```diff
@@ -64,6 +64,73 @@
...
+     <para>The manual tickets are managed by the same commands as normal
+     tickets using the <command>grant</command> or <command>revoke</command>
```
Comment (taroth21) [inline xml/geo_booth_i.xml]: as normal tickets, (insert comma) using the ```grant``` or ```revoke``` subcommands, for example.

File: xml/geo_booth_i.xml [hunk 4]
```diff
@@ -64,6 +64,73 @@
...
+     subcommands.</para>
+    </formalpara>
+   </listitem>
+   <listitem>
+    <formalpara>
+     <title>Unnecessary arbitrator with manual tickets</title>
+     <para>Configuring only manual tickets in a Geo cluster with two sites,
```
Comment (taroth21) [inline xml/geo_booth_i.xml]: Unnecessary arbitrator with manual tickets -> If only manual tickets are used, arbitrators are not needed.

I checked with Chris again (because this was not entirely clear from Fate# 322100): use of manual tickets is not limited to a two-site Geo cluster. Therefore I would rephrase this sentence to: 
If you configure only manual tickets in a Geo cluster, arbitrators are not necessary, because manual ticket management does not require quorum decisions.

File: xml/geo_booth_i.xml [hunk 5]
```diff
@@ -179,6 +242,25 @@ ticket = "&ticket2;" <xref linkend="co.ha.geo.booth.config.ticket" xrefstyle="se
...
+     <para>
+      Optional parameter. Defines how many times booth retries sending packets
+      before giving up waiting for confirmation by other sites. Values smaller
+      than <literal>3</literal> are invalid and will prevent booth from
+      starting.
+     </para>
+    </callout>
+    <callout arearefs="co.ha.geo.booth.mode">
+     <para>Optional parameter. Defines the ticket mode. By default, all
+      tickets are managed by Booth. To define tickets which are managed by
```
Comment (taroth21) [inline xml/geo_booth_i.xml]: Booth -> booth

File: xml/phrases-decl.ent [hunk 6]
```diff
@@ -137,7 +137,7 @@
 ' The IP address of the machine to use as arbitrator. '>
 
  <!ENTITY booth-ticket
- 'The ticket to be managed by booth.'> 
+ 'The tickets to be managed by booth or system administrator.'>
```
Comment (taroth21) [inline xml/phrases-decl.ent]: or system administrator -> or a cluster administrator

File: xml/geo_manage_i.xml [hunk 7]
```diff
@@ -114,6 +127,14 @@ booth[27900]: 2014/08/13_10:21:23 info: revoke succeeded!</screen>
    instead. Find the exact status in the log files or with the
    <command>crm_ticket -L</command> command.
   </para>
+  <para>
+   The <command>grant</command> command can use a <option>-F</option> option.
```
Comment (taroth21) [inline xml/geo_manage_i.xml]: can use -> can take (it's the admin that 'uses' the option) ;)

File: xml/geo_manage_i.xml [hunk 8]
```diff
@@ -114,6 +127,14 @@ booth[27900]: 2014/08/13_10:21:23 info: revoke succeeded!</screen>
...
+   This option forces a site to get a ticket even if the ticket is granted
```
Comment (taroth21) [inline xml/geo_manage_i.xml]: I would rephrase to: 'This option is potentially dangerous, but it can be useful for manual ticket management in case a site which holds a ticket cannot be reached. It makes booth grant the ticket *immediately* to the specified site, even though booth cannot ascertain that other sites (that cannot be reached) do not hold the same ticket.' 

I would also add a warning:
'WARNING Concurrency Violation and Potential Loss of Data
Before you use this option, make sure that no other site currently owns the same ticket. If the same ticket is granted to multiple sites, resources depending on the ticket might be started on multiple sites in parallel, resulting in concurrency violation and potential data corruption.'

**Comment by tomschr** [comment]
@ChrisKowalczyk, @gao-yan
Tanja and I integrated FATE#332100 now. Could you review it? We've integrated some remarks for your where we aren't sure about some issues.

Thanks!

File: xml/geo_booth_i.xml [hunk 9]
```diff
@@ -64,6 +64,75 @@
    parties.
   </para>
  </sect1>
+
+ <sect1 xml:id="sec.ha.geo.booth.ticket-types">
+  <title>Normal versus Manual Tickets</title>
```
Comment (gao-yan) [inline xml/geo_booth_i.xml]: Potentially manual tickets could be more adopted by users. So I wouldn't imply it's abnormal ;-) How about drop the term "normal tickets" and say "Automatic versus Manual Tickets" ?

File: xml/geo_booth_i.xml [hunk 10]
```diff
@@ -104,19 +173,15 @@ site =  192.168.201.100 <co xml:id="co.ha.geo.booth.config.site"/>
 site =  192.168.202.100 <xref linkend="co.ha.geo.booth.config.site" xrefstyle="select:label nopage"/>
 authfile = /etc/booth/authkey <co xml:id="co.ha.geo.booth.config.auth"/>
 ticket = "ticket-nfs" <co xml:id="co.ha.geo.booth.config.ticket"/>
+     retries = 5 <co xml:id="co.ha.geo.booth.config.retries"/>
+     mode = MANUAL <co xml:id="co.ha.geo.booth.mode"/>
```
Comment (gao-yan) [inline xml/geo_booth_i.xml]: @ChrisKowalczyk IIUC "retries" is related to renewal of tickets, and manual tickets don't need to be renewed, probably "retries" doesn't need to be configured here?

File: xml/geo_manage_i.xml [hunk 11]
```diff
@@ -29,8 +29,20 @@
  <sect1 xml:id="sec.ha.geo.manage.cli">
   <title>Managing Tickets From Command Line</title>
 
+  <remark>toms 2018-06-14: there is no "booth client" subcommand. According
```
Comment (gao-yan) [inline xml/geo_manage_i.xml]: I think "client" is basically the default subcommand, meaning unless the special subcommands like "daemon" are supplied,  "booth ..." works same as "booth client", @ChrisKowalczyk?

File: xml/geo_booth_i.xml (hunk 9 above)
Comment (tomschr) [reply xml/geo_booth_i.xml]: Haha, yes. :laughing: The term "automatic" is much better, I've used this.

File: xml/geo_booth_i.xml [hunk 12]
```diff
@@ -64,6 +64,75 @@
...
  </sect1>
+
+ <sect1 xml:id="sec.ha.geo.booth.ticket-types">
+  <title>Normal versus Manual Tickets</title>
+  <para>
+   A ticket grants the right to run certain resources on a specific
+   cluster site. Two types of tickets are supported:</para>
+  <itemizedlist>
+   <listitem>
+    <para>Normal (or automatic) tickets are controlled by the &boothd; daemon.</para>
```
Comment (tomschr) [inline xml/geo_booth_i.xml]: I also removed "normal" here and used the term automatic.

File: xml/geo_manage_i.xml (hunk 11 above)
Comment (tomschr) [reply xml/geo_manage_i.xml]: Hmn, it's a bit confusing. According to the booth man page:

```
SYNOPSIS
       boothd daemon [-SD] [-c config] [-l lockfile]
       booth list [-s site] [-c config]
       booth grant [-s site] [-c config] [-FCw] ticket
       booth revoke [-s site] [-c config] [-w] ticket
       booth peers [-s site] [-c config]
       booth status [-D] [-c config]
```

However, if you look a little bit deeper, I found this:

```
COMMANDS
     Whether the binary is called as boothd or booth doesn’t matter; the first
     argument determines the mode of operation.

       daemon
           Tells boothd to serve a site. The locally configured interfaces are searched for an IP address
           that is defined in the configuration. booth  then runs in either /arbitrator/ or /site/ mode.

       client
           Booth clients can list the ticket information (see also crm_ticket -L), and revoke or grant tickets to a site.
           [...]
```

Well, it could be debated if the man page is inconsistent or structured unfortunate. I assume, to get an overview of all `booth` subcommands you would probably look into the synopsis first. 

Maybe it's just me, but from my point of view, the synopsis of the man page is incomplete.

File: xml/geo_booth_i.xml (hunk 10 above)
Comment (taroth21) [reply xml/geo_booth_i.xml]: Yes, we discussed this already with Chris and a little bit later in the text, it is mentioned  that manual tickets don't have this option. Removed from here with https://github.com/SUSE/doc-sleha/commit/d90d90c3.

**Comment by taroth21** [comment]
PDF of the CLI chapter (with some open questions): 
[cha.ha.geo.manage_remarks_color_draft_en.pdf](https://github.com/SUSE/doc-sleha/files/2133153/cha.ha.geo.manage_remarks_color_draft_en.pdf)

File: xml/geo_manage_i.xml (hunk 11 above)
Comment (gao-yan) [reply xml/geo_manage_i.xml]: The manual is indeed confusing. I guess "client" is not listed in "SYNOPSIS" because it's no longer recommended/needed, given that all the subcommands supported by the client, such are "list", "grant" and so on are listed there already and can be directly invoked. I'd  just skip "client" in our doc. Opinions, @ChrisKowalczyk?

Comment (ChrisKowalczyk) [reply xml/geo_manage_i.xml]: Agreed. I have never used "booth client". Let us skip it!

**Comment by taroth21** [comment]
Updated PDF version of the chapter, including the latest changes:
[cha.ha.geo.manage_remarks_color_draft_en.pdf](https://github.com/SUSE/doc-sleha/files/2137242/cha.ha.geo.manage_remarks_color_draft_en.pdf)
@ChrisKowalczyk @gao-yan : Please review.

File: xml/geo_booth_i.xml [hunk 13]
```diff
@@ -64,6 +64,75 @@
...
  </sect1>
+
+ <sect1 xml:id="sec.ha.geo.booth.ticket-types">
+  <title>Automatic versus Manual Tickets</title>
+  <para>
+   A ticket grants the right to run certain resources on a specific
+   cluster site. Two types of tickets are supported:</para>
+  <itemizedlist>
+   <listitem>
+    <para>Automatic tickets are controlled by the &boothd; daemon.</para>
```
Comment (ChrisKowalczyk) [inline xml/geo_booth_i.xml]: Perhaps, we could add "and RAFT algorithm" or something similar?

File: xml/geo_booth_i.xml [hunk 14]
```diff
@@ -64,6 +64,75 @@
...
+    <formalpara>
+     <title>Normal and manual tickets can be defined together</title>
+     <para>You can define and use both manual and normal tickets within the
+      same Geo cluster.</para>
+    </formalpara>
+   </listitem>
+   <listitem>
+    <formalpara>
+     <title>Manual ticket management remains manual</title>
+     <para>The normal ticket management is not applied to manually controlled
```
Comment (ChrisKowalczyk) [inline xml/geo_booth_i.xml]: use "automatic" instead of "normal"

File: xml/geo_concept_i.xml [hunk 15]
```diff
@@ -228,15 +228,19 @@
    <term>Ticket Failover</term>
    <listitem>
     <para>
-     If the ticket gets lost, which means other booth instances do not hear
-     from the ticket owner in a sufficiently long time, one of the remaining
-     sites will acquire the ticket. This is what is called ticket failover. If
-     the remaining members cannot form a majority, then the ticket cannot fail
-     over (if managed by booth).
+     After you have initially granted a normal ticket to a site, the booth
```
Comment (ChrisKowalczyk) [inline xml/geo_concept_i.xml]: "an automatic" instead of "normal". Also, perhaps "Booth" instead of "the booth mechanism"

File: xml/geo_manage_i.xml [hunk 16]
```diff
@@ -28,161 +28,152 @@
...
    operation's outcome. The client waits for the result up to the ticket's
-   <varname>timeout</varname> value before it gives up waiting. If the
-   <option>-w</option> option was used, the client will wait indefinitely
+    <varname>timeout</varname> value before it gives up waiting. If the
+    <option>-w</option> option was used, the client will wait indefinitely
    instead. Find the exact status in the log files or with the
-   <command>crm_ticket -L</command> command.
-  </para>
+   <command>crm_ticket -L</command> command. </para>
+  <para> The <command>grant</command> command can take a <option>-F</option>
```
Comment (ChrisKowalczyk) [inline xml/geo_manage_i.xml]: Perhaps, worth mentioning here that the meaning of -F is different for automatic tickets?

Comment (taroth21) [reply xml/geo_manage_i.xml]: You mean ```booth grant -F```? Or do you mean ```crm_ticket --force```?

File: xml/geo_booth_i.xml (hunk 13 above)
Comment (taroth21) [reply xml/geo_booth_i.xml]: Hm, I'm hesitant to mention the algorithm. IMHO it's an implementation detail which does not matter to the administrator. In order to focus on the target group of the Geo Clustering Guide (admins), I would prefer not to mention it.

File: xml/geo_manage_i.xml (hunk 16 above)
Comment (gao-yan) [reply xml/geo_manage_i.xml]: "booth grant -F" I think. The behaviors on granting are described in "BOOTH TICKET MANAGEMENT" section from the manual, which is basically for automatics tickets, and "MANUAL TICKETS" section for manual ones of course. Hope they are helpful :-)


  @ChrisKowalczyk?

File: xml/geo_manage_i.xml [hunk 17]
```diff
@@ -28,161 +28,152 @@
...
+     <literal>192.168.201.100</literal>. This site cannot be reached at the
+    moment. Site <literal>&cluster2;</literal> (with the virtual IP
+     <literal>192.168.202.100</literal>) is still available. </para>
+   <para> If you manage to contact a local administrator on
+     <literal>&cluster1;</literal>, you can verify if the site is really down.
+    In that case, the ticket on <literal>&cluster1;</literal> cannot be revoked
+    and the resources depending on this ticket cannot be started on
+     <literal>&cluster2;</literal> unless you manually grant the ticket
+     <literal>ticket-nfs</literal> to <literal>&cluster2;</literal>. To do so,
+    proceed as follows: </para>
```
Comment (gao-yan) [inline xml/geo_manage_i.xml]: How about to rephrase "In that case ...to cluster2" to something like:
"Make sure the ticket, or basically the relevant resources, have been relinquished from cluster1. If we are sure now it's safe and ready to grant the ticket to cluster2 and host the relevant resources there, proceed as follows:"

?

File: xml/geo_manage_i.xml [hunk 18]
```diff
@@ -28,161 +28,152 @@
...
    <step>
-    <para>
-     Revoke <literal>&ticket1;</literal> from site
-     <literal>&cluster1;</literal> with:
-    </para>
-<screen>&prompt.root;<command>booth</command> revoke -s 192.168.201.100 &ticket1;</screen>
+    <para> Before trying to get <literal>&cluster1;</literal> up and running
+     again, set <literal>ticket-nfs</literal> to standby on <literal>&cluster2;</literal> (to
+     prevent that resources will be started on both sites in parallel): </para>
+    <screen>&prompt.root;<command>crm_ticket</command> -t ticket-nfs --standby</screen>
```
Comment (gao-yan) [inline xml/geo_manage_i.xml]: Depending on the needs, if we want to keep the resources running on cluster2, we should not do this since it'd unnecessarily stop the resources and introduce service down time again. I think something like this probably makes more sense:

"Before tying to bring cluster1 back into the geo cluster, again make sure the ticket is revoked and resources are not hosted there."

For the "standby" and "activate" commands, we can introduce them as parts of the normal procedure when doing manual switch-over when all the geo cluster sites are healthy.

File: xml/geo_booth_i.xml (hunk 13 above)
Comment (gao-yan) [reply xml/geo_booth_i.xml]: Agreed with @taroth21

File: xml/geo_manage_i.xml [hunk 19]
```diff
@@ -28,161 +28,152 @@
...
+     to keep the resources running on site <literal>&cluster2;</literal>,
+     revoke <literal>ticket-nfs</literal> from <literal>&cluster1;</literal>: </para>
+    <screen>&prompt.root;<command>booth</command> revoke -s 192.168.201.100 ticket-nfs</screen>
+   </step>
+   <step>
+    <para>Log in to site <literal>&cluster2;</literal> and put
+      <literal>ticket-nfs</literal> into active mode again:</para>
+    <screen>&prompt.root;<command>crm_ticket</command> -t ticket-nfs --activate</screen>
+    <para>All resources that depend on this ticket will be started on
+      <literal>&cluster2;</literal> again. </para>
```
Comment (gao-yan) [inline xml/geo_manage_i.xml]: I can thinking of these potentially interesting management scenarios:

- "Manually move an automatic ticket":
1) put the ticket to standby from cluster1" 
2) "grant the ticket to cluster2" 
3) "activate ticket on cluster1 if we want the resources automatically fail over to cluster1 in case cluster2 fails "
 -- resources run only if the ticket is granted & non-standby.

- "Move a manual ticket"
1) "put the ticket to standby from cluster1"
2) "revoke the ticket from cluster1" 
  -- @Chriskowalczyk If I remember correctly, manual ticket should be explicitly revoked before being granted again, am I correct?
3) "grant the ticket to cluster2" 
4)"activate ticket on cluster1 if we want the resources automatically fail over to cluster1 in case cluster2 fails " 

- "Fail over a manual ticket"
The procedure we already have in here :)

File: xml/geo_manage_i.xml [hunk 20]
```diff
@@ -28,161 +28,152 @@
...
-    </para>
-<screen>&prompt.root;<command>booth</command> grant -s 192.168.202.100 &ticket1;</screen>
+    <para> Wait for any resources that depend on <literal>ticket-nfs</literal>
+     to be stopped or demoted cleanly. </para>
+   </step>
+   <step>
+    <para>Get <literal>&cluster1;</literal> up and running again. If you want
+     to keep the resources running on site <literal>&cluster2;</literal>,
+     revoke <literal>ticket-nfs</literal> from <literal>&cluster1;</literal>: </para>
+    <screen>&prompt.root;<command>booth</command> revoke -s 192.168.201.100 ticket-nfs</screen>
```
[the remark this answers is not in the document]
Comment (ChrisKowalczyk) [reply xml/geo_manage_i.xml]: @gao-yan , hm, I think cluster2 will be confused until it sees revoke..

File: xml/geo_manage_i.xml (hunk 19 above)
Comment (ChrisKowalczyk) [reply xml/geo_manage_i.xml]: @gao-yan , you can grant manual ticket a second time, with -F flag

File: xml/geo_manage_i.xml (hunk 16 above)
Comment (taroth21) [reply xml/geo_manage_i.xml]: @gao-yan , @ChrisKowalczyk: I studied both sections of the man page again and I must admit I don't see much difference with ```booth grant -F``` for manual vs. automatic tickets: if I got it right, in any case the ticket is granted to the new site immediately (without having to wait for  any timeouts or delays - ticket expire time or acquire-after time, which are only defined for automatic tickets, of course). Also, in both cases, it is the responsibility of the admin to make sure the same ticket is not granted to another site already (or that the site who currently has the ticket is really 'dead'). Nevertheless I will try to rephrase and add a separate entry for the ```-F``` option.

Comment (ChrisKowalczyk) [reply xml/geo_manage_i.xml]: @taroth21 , well the meaning is similar in both cases:

* For manual tickets -F means: always grant the ticket even it you know it is granted elsewhere.
* For automatic tickets -F means: grant tickets without waiting for confirmation from other sites. If the application knows that an automatic ticket is granted somewhere, it will not allow to grant it even with -F flag. So, the difference is that for split brain situation, from the app point of view, it may be not clear if an automatic ticket is granted somewhere else. This is why we use -F flag there - "trust me, the ticket is not granted anywhere. do not wait for responses from other, unreachable sites, just grant it"

Comment (taroth21) [reply xml/geo_manage_i.xml]: Thanks, @ChrisKowalczyk, this made it clearer to me. I tried to reflect this in the update of this snippet.

File: xml/geo_manage_i.xml (hunk 17 above)
Comment (taroth21) [reply xml/geo_manage_i.xml]: @gao-yan: But how to make sure the ticket/resources have been relinquished from cluster1 if the site is already down?

**Comment by taroth21** [comment]
@gao-yan @ChrisKowalczyk : Many thanks! I have addressed all your comments but one where I had a question to you: https://github.com/SUSE/doc-sleha/pull/86#discussion_r200401625.

I have attached the updated version of this chapter once more as PDF - please review (hopefully final feedback round on this feature): 

[cha.ha.geo.manage_color_draft_en.pdf](https://github.com/SUSE/doc-sleha/files/2167390/cha.ha.geo.manage_color_draft_en.pdf)

File: xml/geo_booth_i.xml [hunk 21]
```diff
@@ -64,6 +64,75 @@
...
+     <title>Manual ticket management remains manual</title>
+     <para>The automatic ticket management is not applied to manually controlled
+      tickets.
+      Manual tickets do not require any quorum elections, cannot fail over
+      automatically, and do not have an expiry time.</para>
+    </formalpara>
+   </listitem>
+   <listitem>
+    <formalpara>
+     <title>Manual tickets are not moved</title>
```
Comment (ChrisKowalczyk) [inline xml/geo_booth_i.xml]: Perhaps something like "cannot be moved" or "will not be automatically moved" instead of "are not moved" ?

File: xml/geo_booth_i.xml [hunk 22]
```diff
@@ -64,6 +64,75 @@
...
+      the services that depend on a ticket remain on a particular site and
+      are not moved to another site.
+     </para>
+    </formalpara>
+   </listitem>
+   <listitem>
+    <formalpara>
+     <title>Same commands for managing both types of tickets</title>
+     <para>The manual tickets are managed by the same commands as automatic
+     tickets, using the <command>grant</command> or <command>revoke</command>
```
Comment (ChrisKowalczyk) [inline xml/geo_booth_i.xml]: How about changing:
"using the <command>grant</command> or <command>revoke</command> subcommands, for example."
To just:
"for example <command>grant</command> or <command>revoke</command>."

File: xml/geo_booth_i.xml [hunk 23]
```diff
@@ -173,12 +237,23 @@ ticket = "&ticket2;" <xref linkend="co.ha.geo.booth.config.ticket" xrefstyle="se
...
-      For example, the ticket <literal>ticket-nfs</literal> specified here can
+      For example, the ticket <literal>&ticket3;</literal> specified here can
       be used for failover of NFS and DRBD as explained in
       <link
        xlink:href="&suse-onlinedoc;suse-best-practices/sbp-drbd/data/sbp-drbd.html"/>.
      </para>
     </callout>
+   <callout arearefs="co.ha.geo.booth.mode">
+     <para>Optional parameter. Defines the ticket mode. By default, all
+      tickets are managed by booth. To define tickets which are managed by
```
Comment (ChrisKowalczyk) [inline xml/geo_booth_i.xml]: "tickets are managed by booth"  ->  "tickets are managed automatically by booth"

File: xml/geo_concept_i.xml [hunk 24]
```diff
@@ -228,15 +228,19 @@
    <term>Ticket Failover</term>
    <listitem>
     <para>
-     If the ticket gets lost, which means other booth instances do not hear
-     from the ticket owner in a sufficiently long time, one of the remaining
-     sites will acquire the ticket. This is what is called ticket failover. If
-     the remaining members cannot form a majority, then the ticket cannot fail
-     over (if managed by booth).
+     After you have initially granted an automatic ticket to a site, booth
+     takes over and manages the ticket automatically.
```
Comment (ChrisKowalczyk) [inline xml/geo_concept_i.xml]: "booth takes over and manages the ticket automatically"  ->  "booth will manage this ticket automatically"  ?

File: xml/geo_concept_i.xml [hunk 25]
```diff
@@ -228,15 +228,19 @@
...
+     If the site holding a ticket should be out of service, the ticket is
+     automatically revoked after the expiry time.
+     If the remaining sites have quorum, the ticket will be granted to another
+     site (fail over). The resources that depend on that ticket fail over to
+     the new site that holds the ticket. The <literal>loss-policy</literal>
+     set within the constraint specifies what happens to the nodes that have
+     run the resources before.
```
Comment (ChrisKowalczyk) [inline xml/geo_concept_i.xml]: "The loss-policy set within the constraint[...]"  ->  this is not completely clear for me.. Perhaps, we could clarify this a little, like: "The loss-policy, which was defined according to the .... , ..." ?

File: xml/geo_manage_i.xml [hunk 26]
```diff
@@ -28,130 +28,144 @@
...
+     </itemizedlist>
+     <warning>
+      <title>Potential Loss of Data</title>
+      <para>Before using <command>booth grant -F</command>, make sure that no
+       other site (which is online) owns the same ticket. If the same ticket is
+       granted to multiple sites, resources depending on the ticket might start
+       on several sites in parallel. This results in concurrency violation and
+       potential data corruption.</para>
+      <para>As &geo; cluster administrator, you need to resolve a ticket
+       conflict once the other site is reachable again.</para>
```
Comment (ChrisKowalczyk) [inline xml/geo_manage_i.xml]: "a ticket conflict"  ->  "the conflict between tickets"   ?

File: xml/geo_manage_i.xml [hunk 27]
```diff
@@ -28,130 +28,144 @@
...
-   that ticket fail over to the new site that holds the ticket. The
-   <literal>loss-policy</literal> set within the constraint specifies what
-   happens to the nodes that have run the resources before.
-  </para>
-
-<!--taroth 2013-04-24: fix for bnc#752601, c#10-->
-
-  <procedure xml:id="pro.ha.geo.manage.tickets">
-   <title>Managing Tickets Manually</title>
+  <para>In the following, find some examples for managing tickets in different
```
Comment (ChrisKowalczyk) [inline xml/geo_manage_i.xml]: "In the following"  ?  maybe better "find some example  below" ? or "in the next paragraph"

File: xml/geo_manage_i.xml [hunk 28]
```diff
@@ -27,131 +27,164 @@
...
-   will always connect to the local site.
+  <title>Managing Tickets From Command Line</title>
+  <para> Use the <command>booth</command> command line tool to grant, list, or
+   revoke tickets as described in <xref linkend="sec.ha.geo.manage.cli.booth"/>.
   </para>
+  <warning>
+   <title><command>crm_ticket</command> and
+    <command>crm&nbsp;site&nbsp;ticket</command></title>
+   <para> If the booth service is not running for any reasons, you can also
+    manage tickets manually with <command>crm_ticket</command> or
```
Comment (gao-yan) [inline xml/geo_manage_i.xml]: How about "purely manually" or "fully manually", to differentiate from the management of manual tickets?

File: xml/geo_manage_i.xml [hunk 29]
```diff
@@ -27,131 +27,164 @@
...
+depending on this ticket cannot be started on berlin unless you manually grant the ticket
+ticket-nfs to berlin.
+
+ygao 2018-06-26: @taroth:How about to rephrase "In that case ...to cluster2" to be something like:
+"Make sure the ticket, or basically the relevant resources, have been relinquished
+ from cluster1. If we are sure now it's safe and ready to grant the ticket to cluster2
+ and host the relevant resources there, proceed as follows:"
+
+taroth 2018-07-03: @gyan: But how to make sure the ticket/resources have been
+relinquished from cluster1 if the site is already down?
```
Comment (gao-yan) [inline xml/geo_manage_i.xml]: By contacting the administrators over there with telephone/email/letter ;-)

File: xml/geo_manage_i.xml [hunk 30]
```diff
@@ -164,25 +197,115 @@ booth[27900]: 2014/08/13_10:21:23 info: revoke succeeded!</screen>
...
+     <!--After the ticket has been revoked from its original site, -->Grant
+     <literal>&ticket1;</literal> to the site <literal>&cluster2;</literal> with:
     </para>
 <screen>&prompt.root;<command>booth</command> grant -s 192.168.202.100 &ticket1;</screen>
    </step>
+   <step>
+    <para>
+     Activate <literal>&ticket1;</literal> on site <literal>&cluster1;</literal> with:
+    </para>
+<screen>&prompt.root;<command>booth</command> grant -a 192.168.202.100 &ticket1;</screen>
```
Comment (gao-yan) [inline xml/geo_manage_i.xml]: This needs to be done by "crm_ticket -t ticket1 -a" in cluster1.

File: xml/geo_manage_i.xml [hunk 31]
```diff
@@ -164,25 +197,115 @@ booth[27900]: 2014/08/13_10:21:23 info: revoke succeeded!</screen>
...
-  <para>
-   <remark>taroth 2018-04-03: FATE#322100 - add how to proceed for manual failover
-    of tickets</remark>
-  </para>
- </sect1>
+ </sect2>
+ <sect2 xml:id="sec.ha.geo.manage.cli.booth.man-ticket.move">
+  <title>Moving a Manual Ticket</title>
+   <para>
+    Assuming that you want to manually move <literal>ticketnfs</literal> from
```
Comment (gao-yan) [inline xml/geo_manage_i.xml]: How about "move the manual ticket 'ticketnfs'"? Since "manually" is the only way ;-)

File: xml/geo_manage_i.xml [hunk 32]
```diff
@@ -164,25 +197,115 @@ booth[27900]: 2014/08/13_10:21:23 info: revoke succeeded!</screen>
...
    </step>
-   <step>
+  <!-- <step>
     <para>
      Revoke <literal>&ticket1;</literal> from site
      <literal>&cluster1;</literal> with:
     </para>
 <screen>&prompt.root;<command>booth</command> revoke -s 192.168.201.100 &ticket1;</screen>
-   </step>
+   </step>-->
```
Comment (gao-yan) [inline xml/geo_manage_i.xml]: We need to explicitly revoke the ticket  first before granting it again. @ChrisKowalczyk confirmed so :)

File: xml/geo_manage_i.xml [hunk 33]
```diff
@@ -164,25 +197,115 @@ booth[27900]: 2014/08/13_10:21:23 info: revoke succeeded!</screen>
...
+    <para>
+     Wait for any resources that depend on <literal>&ticket3;</literal> to be
+     stopped or demoted cleanly.
+    </para>
+   </step>
+   <step>
+    <para>
+     Grant <literal>&ticket3;</literal>to the site <literal>&cluster2;</literal> with:
+    </para>
+<screen>&prompt.root;<command>booth</command> grant -s 192.168.202.100 -F &ticket3;</screen>
```
Comment (gao-yan) [inline xml/geo_manage_i.xml]: Similarly, we should explicitly revoke the ticket first and then grant it again without "-F". Unnecessary "-F" would introduce double-granting (concurrency violation).

File: xml/geo_manage_i.xml [hunk 34]
```diff
@@ -164,25 +197,115 @@ booth[27900]: 2014/08/13_10:21:23 info: revoke succeeded!</screen>
...
+   </step>
+   <step>
+    <para>
+     Activate <literal>&ticket3;</literal> on site <literal>&cluster1;</literal> with:
+    </para>
+<screen>&prompt.root;<command>booth</command> grant -a 192.168.202.100 &ticket1;</screen>
+     <para>Resources depending on <literal>&ticket3;</literal> will automatically fail over to
+       <literal>&cluster1;</literal> in case
+       <literal>&cluster2;</literal> fails.</para>
```
[the remark this answers is not in the document]
Comment (gao-yan) [reply xml/geo_manage_i.xml]: So it seems like this scenario is quite similar to the previous one now ;-)

File: xml/geo_manage_i.xml [hunk 35]
```diff
@@ -164,25 +197,115 @@ booth[27900]: 2014/08/13_10:21:23 info: revoke succeeded!</screen>
...
+     <literal>&cluster1;</literal> (with the virtual IP
+     <literal>192.168.201.100</literal>. This site cannot be reached at the
+    moment. Site <literal>&cluster2;</literal> (with the virtual IP
+     <literal>192.168.202.100</literal>) is still available. </para>
+   <para> If you manage to contact a local administrator on
+     <literal>&cluster1;</literal>, you can verify if the site is really down.
+    In that case, the ticket on <literal>&cluster1;</literal> cannot be revoked
+    and the resources depending on this ticket cannot be started on
+     <literal>&cluster2;</literal> unless you manually grant the ticket
+     <literal>&ticket3;</literal> to <literal>&cluster2;</literal>. To do so,
```
Comment (gao-yan) [inline xml/geo_manage_i.xml]: The comment applies in here now ;-)
```
ygao 2018-06-26: @taroth:How about to rephrase "In that case ...to cluster2" to be something like:
"Make sure the ticket, or basically the relevant resources, have been relinquished
from cluster1. If we are sure now it's safe and ready to grant the ticket to cluster2
and host the relevant resources there, proceed as follows:"
```

File: xml/geo_manage_i.xml (hunk 30 above)
Comment (taroth21) [reply xml/geo_manage_i.xml]: Thanks, changed.

File: xml/geo_manage_i.xml (hunk 31 above)
Comment (taroth21) [reply xml/geo_manage_i.xml]: Right :)

File: xml/geo_manage_i.xml (hunk 32 above)
Comment (taroth21) [reply xml/geo_manage_i.xml]: OK, changed.

File: xml/geo_manage_i.xml (hunk 34 above)
[the remark this answers is not in the document]
Comment (ChrisKowalczyk) [reply xml/geo_manage_i.xml]: @gao-yan, these comments are not completely clear to me.  Are you sure about this "booth grant -a" activate command, I have never used it before?

File: xml/geo_manage_i.xml (hunk 35 above)
Comment (gao-yan) [reply xml/geo_manage_i.xml]: Cluster1 cannot be reached for a reason. It might be really down, which is not bad :-) -- We can do the move right away. It might not be really down though, for example if there's network connectivity issue. If it's latter, and if we want to switch things to cluster2, in order to prevent any concurrency violations, we should still ask admin at cluster1 to revoke/standby the ticket over there by itself,  so that the cluster/site basically relinquishes the resources,

File: xml/geo_manage_i.xml (hunk 34 above)
[the remark this answers is not in the document]
Comment (gao-yan) [reply xml/geo_manage_i.xml]: I meant the description "automatically fail over" in here won't happen for manual tickets.

Sooner or later, when we want to move the manual ticket back to cluster1, we of course need to take the ticket out of standby mode, so that  resources are actually hosted,

[the remark this answers is not in the document]
Comment (gao-yan) [reply xml/geo_manage_i.xml]: As commented in previous section -- "Manually Moving an Automatic Ticket", taking a ticket out of standby mode needs to be done with "crm_ticket -t ticket1 -a" .
